### PR TITLE
refactor() the overall codebase refactor WIP

### PR DESCRIPTION
**Pull request by kamilmysliwiec**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[x] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[x] Build related changes
[ ] CI related changes
[x] Other... Please describe: major release
```

## What is the current behavior?

Issue Number: N/A

### Changes/features

- the overall codebase refactor (improve typings, extract common functionalities)
- add tests (+ validate schema against OpenAPI validator)
- migrate to OpenAPI 3.0
- add security-related decorators (e.g. `@ApiSecurity()`, `@ApiBasicAuth()`)
- allow specifying schema manually for more sophisticated scenarios (e.g. nested arrays)
- breaking changes:

1. All `@ApiImplicit{...}` decorators have been changed to `@Api{...}`, for example, `@ApiImplicitBody()` -> `@ApiBody()` - reason: the name was confusing because this operation actually wasn't "implicit".
2. Both `@ApiResponseProperty()` and `@ApiModelProperty` -> `@ApiProperty()`.

### Fixes

- https://github.com/nestjs/swagger/issues/350
- https://github.com/nestjs/swagger/issues/342
- https://github.com/nestjs/swagger/issues/341
- https://github.com/nestjs/swagger/issues/324
- https://github.com/nestjs/swagger/issues/315
- https://github.com/nestjs/swagger/issues/312
- https://github.com/nestjs/swagger/issues/311
- https://github.com/nestjs/swagger/issues/275
- https://github.com/nestjs/swagger/issues/257
- https://github.com/nestjs/swagger/issues/250
- https://github.com/nestjs/swagger/issues/245
- https://github.com/nestjs/swagger/issues/237
- https://github.com/nestjs/swagger/issues/192
- https://github.com/nestjs/swagger/issues/191
- https://github.com/nestjs/swagger/issues/185
- https://github.com/nestjs/swagger/issues/181
- https://github.com/nestjs/swagger/issues/176
- https://github.com/nestjs/swagger/issues/172
- https://github.com/nestjs/swagger/issues/147
- https://github.com/nestjs/swagger/issues/139
- https://github.com/nestjs/swagger/issues/140
- https://github.com/nestjs/swagger/issues/130
- https://github.com/nestjs/swagger/issues/83
- https://github.com/nestjs/swagger/issues/63

### Future

As I've mentioned in this issue https://github.com/nestjs/swagger/issues/191, I planned to rewrite the entire package to use AST and generate JSON schema as an additional compilation step. However, this change would make the usage of the package slightly more complicated. In addition, most of the code written so far would have to be removed. Hence, I came up with another solution.

The entire package has been somewhat refactored & most of the common issues should be addressed within this PR. Now, the next step would be to create a **plugin**.

As you can read in this https://trilon.io/blog/announcing-nestjs-monorepos-and-new-commands blog post, new CLI features introduced plugins system. Thus, we can easily use the existing package and add some new features on top of it based on the AST and pre-compilation hook.

The plugin would allow us to:
- generate the default responses (based on the return type - extracted either `Promise` or `Observable`)
- automatically add `@ApiProperty()` if not present
- automatically set the property to either `required: true` or `required: false` depending on the question mark `?` (optional value)
- automatically transform arrays, for example
```typescript
before:
----
@ApiProperty()
tags: string[]

after (this step is automated - "behind the curtain"):
----
@ApiProperty({ type: [String] })
tags: string[]
```
- conditionally (if enabled) merge `class-validator` validation rules with OpenAPI properties
- support generics (and maybe interfaces?, i'm not sure about the potential use-case yet)

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

## Other information

### CODE REVIEW

**Comment by GillesWHC** [comment]
This refactor fixes my greatest gripes with the module. Great work!

**Comment by micaelmbagira** [comment]
Awesome! @kamilmysliwiec I liked the `generate JSON schema as an additional compilation step`, in my use case, it's convenient to extract the JSON or YAML file during a build step.
Once the plugin is implemented, this JSON generation would be trivial, am I correct?

**Comment by nartc** [comment]
@kamilmysliwiec Looks great!! Is there a way to contribute to `refactor/major` branch? Or you want to roll out a `beta` version before allowing contribution? Thanks for the hard work 💪

**Comment by ulrichb** [comment]
@kamilmysliwiec Because it's not explicitly listed, will this also fix #116?

**Comment by kamilmysliwiec** [comment]
`@nestjs/swagger@4.0.0-next.1` under `@next` tag has just been published (this PR)

**Comment by johannesschobel** [comment]
Dear @kamilmysliwiec , thanks a lot for this update! Does this new version include "open api v3"? i would totally love to try it out - i'm quite hyped for this!

**Comment by kamilmysliwiec** [comment]
### Breaking Changes

1. The following decorators have been changed:
- `@ApiModelProperty` -> `@ApiProperty`
- `@ApiModelPropertyOptional` -> `@ApiPropertyOptional`
- `@ApiImplicitQuery` -> `@ApiQuery`
- `@ApiImplicitParam` -> `@ApiParam`
- `@ApiImplicitBody` -> `@ApiBody`
- `@ApiImplicitHeader` -> `@ApiHeader`
- `@ApiOperation({ title: 'test' })` ->`@ApiOperation({ summary: 'test' })`

2. `DocumentBuilder` breaking changes (aka updated methods signatures):

- `addTag`
- `addBearerAuth`
- `addOAuth2`
- `setContactEmail` -> `setContact` (the entire contact information object)
- `setHost` has been removed
- `setSchemes` has been removed

+ the following methods have been added:

- `addServer`
- `addApiKey`
- `addBasicAuth`
- `addSecurity`
- `addSecurityRequirements`

**Comment by kamilmysliwiec** [comment]
@johannesschobel yes, OpenAPI 3.0 is supported

**Comment by johannesschobel** [comment]
Just wanted to give a quick feedback on the OA3 integration.. Already started working on integrating the new version!

## this is huge! i absolutely love it! :heart:

Minor suggestion though:
I would rename the `@ApiUseTags(...string)` to `ApiTags(...string)` in order to better align this decorator as well.. What do you think @kamilmysliwiec  ?

**Comment by nartc** [comment]
> Just wanted to give a quick feedback on the OA3 integration.. Already started working on integrating the new version!
> 
> ## this is huge! i absolutely love it! ❤️
> Minor suggestion though:
> I would rename the `@ApiUseTags(...string)` to `ApiTags(...string)` in order to better align this decorator as well.. What do you think @kamilmysliwiec ?

I would agree.

**Comment by kamilmysliwiec** [comment]
Sounds good @nartc  @johannesschobel. I'll do it shortly

**Comment by nartc** [comment]
@kamilmysliwiec I'll start working on the docs.

**Comment by kamilmysliwiec** [comment]
I'm actually working on the docs locally already, but any additional suggestions would be very useful. :D

**Comment by nartc** [comment]
Then I'll wait and make suggestive changes later as we see fit :D Kudos to you man.

**Comment by loban** [comment]
I've either found a bug or am incorrectly using this version. I changed all cases of `type: TYPE` to `type: () => TYPE` in `@ApiOperation` and `@ApiProperty` decorators. Unfortunately, after doing this, the Swagger Explorer does not show output schemas at all.

**Comment by nartc** [comment]
@loban check #369

**Comment by loban** [comment]
@nartc Unfortunately your branch did not fix my issue.

**Comment by nartc** [comment]
It fixed it for me. You have a repro?

**Comment by loban** [comment]
> It fixed it for me. You have a repro?

It's a private project. Lemme first double-check to make sure I didn't screw up pulling your branch. And if I still have a problem. I'll try to whip up a public repo that show-cases it.

**Comment by loban** [comment]
> > It fixed it for me. You have a repro?
> 
> It's a private project. Lemme first double-check to make sure I didn't screw up pulling your branch. And if I still have a problem. I'll try to whip up a public repo that show-cases it.

Ok the latest 4.0.0-next.3 still has my bug. I'm gonna whip up an example and issue a ticket in a bit.

**Comment by nartc** [comment]
Sounds good!

**Comment by loban** [comment]
> Sounds good!

I've posted an issue with a link to my fork, in which I modified the test case to expose the error.

https://github.com/nestjs/swagger/issues/374

**Comment by loban** [comment]
> > Sounds good!
> 
> I've posted an issue with a link to my fork, in which I modified a test case to expose the error.
> 
> #374

Made a PR to fix at #375

**Comment by klangenk** [comment]
> * conditionally (if enabled) merge `class-validator` validation rules with OpenAPI properties

@kamilmysliwiec what exactly does this point mean? It would be cool to get documentation and validation with one decorator.

**Comment by kamilmysliwiec** [comment]
> @kamilmysliwiec what exactly does this point mean? It would be cool to get documentation and validation with one decorator.

It will be possible to pass configuration object to the plugin (to either disable or enable this behavior)

**Comment by tocosastalo** [comment]
@kamilmysliwiec and guys great job 👍 

We experiment a little with the version `"@nestjs/swagger": "^4.0.0-next.11"` and have problem with validity of outputed scheme.
```
// CompanyDto
export class CompanyDtoGet {
  @ApiProperty({ description: 'id' })
  readonly id: string;

  @ApiProperty({ description: 'name' })
  readonly name: string;
}

// UseDto
export class UserDtoGet {
  @ApiProperty({ description: 'user company' })
  readonly company: CompanyDtoGet;
}
```

is converted into:
```
components:
 schemas:
  UserDtoGet:
       type: object
       properties:
         id:
           type: string
           description: id
        ...
         company:
           title: CompanyDtoGet
           schema: <-- Structural error at components.schemas.UserDtoGet.properties.company
             allOf:
               - $ref: '#/components/schemas/CompanyDtoGet'
               - description: user company
```

and it throws error: 
```
should NOT have additional properties additionalProperty: schema
```

When I remove `schema` key, it seems to be valid. Any suggestions? 
Thanks

**Comment by kamilmysliwiec** [comment]
@tocosastalo thanks for reporting. Can you test 4.0.0-next.12?

**Comment by tocosastalo** [comment]
@kamilmysliwiec well done. Looks good, thanks.

And any updates about `oneOf`, etc. ... https://github.com/nestjs/swagger/pull/355#issuecomment-544476645 ?

**Comment by ulrichb** [comment]
@kamilmysliwiec Is https://github.com/nestjs/swagger/pull/355#issuecomment-547925879 still up to date? Or is there already some kind of migration guide doc?

**Comment by kamilmysliwiec** [comment]
@ulrichb you can take a look at this PR https://github.com/nestjs/docs.nestjs.com/pull/825 (**migration** section)

**Comment by ulrichb** [comment]
@kamilmysliwiec ❤️

**Comment by ulrichb** [comment]
@kamilmysliwiec `.setBasePath()` has been removed, but https://github.com/nestjs/swagger/issues/105 is still not working (tested with _4.0.0-next.12_). => The generated swagger does not prefix the route paths :/

**Comment by johannesschobel** [comment]
try adding the configuration like so:
```
servers: [
        {
          description: 'Local Development Server',
          url: 'http://localhost:3000/api',
        },
      ],
```

**Comment by ulrichb** [comment]
@johannesschobel Thanks! (I'm using the new builder method `.addServer("/my-prefix")`.)

**Comment by johannesschobel** [comment]
@ulrichb i have most of my swagger-configuration in a dedicated `swagger.config.ts` file and just load them like so:
```
const swaggerDocument = SwaggerModule.createDocument(
      app,
      config.swagger.configuration as OpenAPIObject,
      config.swagger.options,
    );
    SwaggerModule.setup('docs', app, swaggerDocument);
```

The `config.swagger.configuration` uses the `OpenAPIObject` from `@nestjs/swagger` which provides the required interface structures

**Comment by sunada-ke** [comment]
Thanks for Open API 3.0 support!!

**Comment by ratio91** [comment]
Awesome!

However, I get this warning:
`UnhandledPromiseRejectionWarning: TypeError: httpAdapter.getType is not a function
    at Function.setup (/usr/src/app/node_modules/@nestjs/swagger/dist/swagger-module.js:14:40)`

My document builder looks like this (main.ts):
```  
const options = new DocumentBuilder()
    .setTitle('API')
    .addTag('agent')
    .addTag('entity')
    .addTag('file')
    .addBearerAuth()
    .setVersion('1.0')
    .addServer('/api-docs')
    .build()
  const document = SwaggerModule.createDocument(app, options)
  SwaggerModule.setup('/api-docs', app, document, {customSiteTitle:'API documentation'})
```

I updated to the latest swagger & nest & TS version.
Any idea what could be wrong? 
Thank you!

**Comment by kamilmysliwiec** [comment]
@ratio91 update `@nestjs/platform-express`

**Comment by ulrichb** [comment]
Since next-13 (or 14), the "GlobalPrefix" is now _automatically_ part of (all) the operation paths, instead of the "server" part (using `.addServer("/myprefix")`) of the output document. Is this an intended change?

For me that's a problem, because now my NSwag generated client code contains the global prefix "hard coded" in all operations, and I cannot influence the prefix via the "base URL" anymore. (And I have varying prefixes.)

In other words: My problem is that the global prefix (which can vary in my case) is now a "hard coded" part of the swagger operation paths, which conflicts with by swagger client tooling.

If this change is intended, could we add an option to disable it?

**Comment by kamilmysliwiec** [comment]
You can use `ignoreGlobalPrefix` (`createDocument` method, 3rd argument)

**Comment by ulrichb** [comment]
Okay, cool thanks for the option. But I think atm. (next 15) it gets applied reversed (`ignoreGlobalPrefix: true` => I see the prefix in the op paths; `ignoreGlobalPrefix: false` => no prefixes).

**Comment by kamilmysliwiec** [comment]
Thanks for the quick feedback. Just pushed another release @ulrichb.

@ulrichb Did you give a shot to `@nestjs/swagger/plugin` too?

**Comment by ulrichb** [comment]
@kamilmysliwiec: Yeah, now it works great!

Regarding swagger/plugin: Sorry (up to now) I don't use Nest CLI at all and neither webpack; just plain old `tsc`.

**Comment by mhendlich** [comment]
There seems to be an issue with the typings when running in `strict ` ts checking mode:
`node_modules/@nestjs/swagger/dist/interfaces/open-api-spec.interface.d.ts:123:5 - error TS2411: Property 'default' of type 'ReferenceObject | ResponseObject | undefined' is not assignable to string index type 'ReferenceObject | ResponseObject'.`

[Repo to reproduce](https://github.com/mhendlich/swaggerbug) (added newest swagger@next, added cli plugin, enabled strict tsc). Run `start:dev` and check outpout.

**Comment by ulrichb** [comment]
@kamilmysliwiec BTW: We could use the major version bump to fix https://github.com/nestjs/swagger/issues/116 ...

**Comment by kamilmysliwiec** [comment]
@mhendlich you can use `skipLibCheck: true` in `tsconfig.json`

@ulrichb done (update to the latest version)

**Comment by ulrichb** [comment]
> @ulrichb done (update to the latest version)

Nice! Many thanks!

**Comment by mhendlich** [comment]
There seems to be an issue with the "classValidatorShim" option of the plugin:
If the DTO file isn't in the same folder as the controller, some paths cannot be resolved when the plugin is running. 

`internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module './dto        est.dto'`

I updated [my demo repository](https://github.com/mhendlich/swaggerbug) for this bug. If you move all files from the dto folder into src, it works fine.

**Comment by kamilmysliwiec** [comment]
@mhendlich I just tried with your repo and I can't reproduce your issue. What's your OS? Could you share any screenshots, push `dist` folder (compilation output) to repo as well?

**Comment by mhendlich** [comment]
> 
> 
> @mhendlich I just tried with your repo and I can't reproduce your issue. What's your OS? Could you share any screenshots, push `dist` folder (compilation output) to repo as well?

Im running Win10. It reproduces on another machine, but seems to work fine under Linux (tried on WSL/Ubuntu). I added both dist folders to the repository.
Node version is 13.2.0 in both cases.

**Comment by kamilmysliwiec** [comment]
Thanks @mhendlich! Fixed in @nestjs/swagger@4.0.0-next.19

**Comment by klangenk** [comment]
Is there already a way to generate some of the class-validator decorators (IsDefined, IsEnum, etc.) in the same way like the swagger plugin does?

**Comment by mkolbusz** [comment]
Should `@ApiResponseProperty()` add request parameter? I think it should not. When I add this decorator to a class property I can see in the generated swagger documentation the field in response as well in the request body as a required parameter. I need to generate class property only in response and ignore in a request. Is it possible?

**Comment by mklan** [comment]
Hello I have seen that the custom redirect url feature [#324 ](https://github.com/nestjs/swagger/issues/324) was merged, but I cannot figure out how to use it. 

My problem is that when I try to authorize trough oauth2 the redirect url gets set to http://localhost:3000/oauth2-redirect.html instead of http://localhost:3000/api/oauth2-redirect.html

@HansG89 have you figured it out? Can somebody point me in the right direction? Thank you very much.

**Comment by mklan** [comment]
Ok I solved it:

```javascript
  SwaggerModule.setup('/api', app, document, {
    swaggerOptions: {
      oauth2RedirectUrl: `http://localhost:${port}/api/oauth2-redirect.html`,
      oauth: {
        clientId: process.env.SWAGGER_OAUTH2_CLIENT_ID || '',
      },
    },
  });
```

**Comment by alexandrulesi** [comment]
> > @kamilmysliwiec what exactly does this point mean? It would be cool to get documentation and validation with one decorator.
> 
> It will be possible to pass configuration object to the plugin (to either disable or enable this behavior)

Have there been any developments on this?
